File: LeetCode/101-200/169. Majority Element/Solution.cpp
/* Boyer Moore Algorithm*/
// Time Complexity: O(n)
// Space Complexity: O(1)
class Solution {
public:
  int majorityElement(vector<int> &nums) {
    int ans = 0;
    int count = 0;
    for (int i = 0; i < nums.size(); i++) {
      if (count == 0)
        ans = nums[i];
      if (nums[i] == ans)
        count++;
      else
        count--;
    }
    return ans;
  }
};

// Time Complexity: O(nlogn)
class Solution {
public:
  int majorityElement(vector<int> &nums) {
    sort(nums.begin(), nums.end());
    return nums[nums.size() / 2];
  }
};

File: LeetCode/101-200/169. Majority Element/Solution.py
# Boyer Moore Algorithm
# Time COmplexity: O(N)
# Space COmplexity: O(1)
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        ans, count = 0, 0
        for n in nums:
            if count == 0:
                ans = n
            count += (1 if ans == n else -1)
        return ans


# Time COmplexity: O(N)
# Space COmplexity: O(N)
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        count = {}
        ans, maxCount = 0, 0
        for n in nums:
            count[n] = 1 + count.get(n, 0)
            ans = n if count[n] > maxCount else ans
            maxCount = max(count[n], maxCount)
        return ans